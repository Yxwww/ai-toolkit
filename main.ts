import Anthropic from "npm:@anthropic-ai/sdk";

const anthropic = new Anthropic();

const msg = await anthropic.messages.create({
  model: "claude-3-7-sonnet-20250219",
  max_tokens: 1000,
  temperature: 1,
  system: "Respond only with short poems.",
  messages: [
    {
      role: "user",
      content: [
        {
          type: "text",
          text: "Why is the ocean salty?",
        },
      ],
    },
  ],
});
console.log(msg);
