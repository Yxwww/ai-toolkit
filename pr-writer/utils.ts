const textDecoder = new TextDecoder();

export const readPrTemplate = async () => {
  try {
    // First try to read from .github/pull_request_template.md
    const prTemplatePath = ".github/pull_request_template.md";
    let prTemplate = "";
    try {
      prTemplate = await Deno.readTextFile(prTemplatePath);
      console.log(`Read PR template from ${prTemplatePath}`);
      return prTemplate;
    } catch (error) {
      // If that fails, try to read from .github/PULL_REQUEST_TEMPLATE folder
      try {
        const prTemplateDirPath = ".github/PULL_REQUEST_TEMPLATE";
        const files = [...Deno.readDirSync(prTemplateDirPath)].filter((file) =>
          file.isFile &&
          (file.name.endsWith(".md") || file.name.endsWith(".txt"))
        );

        if (files.length > 0) {
          // Use the first template file found
          prTemplate = await Deno.readTextFile(
            `${prTemplateDirPath}/${files[0].name}`,
          );
          console.log(
            `Read PR template from ${prTemplateDirPath}/${files[0].name}`,
          );
          return prTemplate;
        }
      } catch (nestedError) {
        // Both attempts failed
        console.log(
          "Could not find PR template in .github directory. Use Deafult Template instead.",
        );
        return prTemplate;
      }
    }
  } catch (error) {
    console.error(
      "Error reading PR template. Fallback to Default Template. Error:",
      error,
    );
    return DEFAULT_PR_TEMPLATE;
  }

  return DEFAULT_PR_TEMPLATE;
};

export const readLastCommit = async () => {
  try {
    const message = await getLatestCommitMessage();
    const diffSummary = await getDiffSummary();
    // Get the actual code changes from the last commit
    const process = new Deno.Command("git", {
      args: ["show", "--pretty=format:", "--patch"],
      stdout: "piped",
      stderr: "piped",
    });

    const output = await process.output();
    const codeChanges = textDecoder.decode(output.stdout).trim();

    const commit = `
Commit Message:
${message}

Changes:
${diffSummary}

Code Changes:
${codeChanges}
    `.trim();

    console.log("Read last commit");
    return commit;
  } catch (error) {
    console.error("Error reading last commit:", error);
    return "";
  }
};

const getLatestCommitMessage = async () => {
  try {
    const process = new Deno.Command("git", {
      args: ["log", "-1", "--pretty=%B"],
      stdout: "piped",
      stderr: "piped",
    });

    const output = await process.output();
    const result = textDecoder.decode(output.stdout).trim();

    return result;
  } catch (error) {
    console.error("Error reading latest commit message:", error);
    return "";
  }
};
const getDiffSummary = async () => {
  try {
    // Get the diff between HEAD and the previous commit to show the latest changes
    const process = new Deno.Command("git", {
      args: ["diff", "HEAD^", "HEAD", "--stat"],
      stdout: "piped",
      stderr: "piped",
    });

    const output = await process.output();
    const diffSummary = new TextDecoder().decode(output.stdout).trim();
    console.log("Latest diff summary:", diffSummary);

    return diffSummary;
  } catch (error) {
    console.error("Error getting diff summary:", error);
    return "";
  }
};

const DEFAULT_PR_TEMPLATE = `
  ## PR Description
  <!-- Provide a brief summary of the changes -->

  ## Changes Made
  <!-- List the main changes introduced by this PR -->

  ## Testing
  <!-- Describe how you tested these changes -->

  ## Related Issues
  <!-- Reference any related issues: "Fixes #123" or "Related to #456" -->

  ## Screenshots (if applicable)
  <!-- Add any relevant screenshots -->

  ## Additional Notes
  <!-- Any other information that might be helpful -->
  `;
