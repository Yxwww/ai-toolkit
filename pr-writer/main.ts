import Anthropic from "npm:@anthropic-ai/sdk";
import { parseArgs } from "jsr:@std/cli/parse-args";
import { readLastCommit, readPrTemplate } from "./utils.ts";

// Helper to mock anthropic in tests
const {
  getAnthropic,
  _mockAnthropic,
} = (() => {
  let anthropic: Anthropic;
  return {
    getAnthropic: () => {
      if (anthropic) return anthropic;
      anthropic = new Anthropic();
      return anthropic;
    },
    _mockAnthropic: (mockClient: unknown) => {
      const original = anthropic;
      anthropic = mockClient as typeof anthropic;
      return original;
    },
  };
})();

const isTestMode = Deno.env.get("test") === "true";
// Export for testing
export async function main() {
  const args = parseArgs(Deno.args);
  console.log("args", args);

  const anthropic = getAnthropic();
  const textEncoder = new TextEncoder();

  const template = await readPrTemplate();
  const lastCommit = await readLastCommit();
  if (lastCommit.length > 10000) {
    throw new Error("Commit message is too long");
  }
  const msg = await anthropic.messages.create({
    model: "claude-3-7-sonnet-20250219",
    max_tokens: 1000,
    temperature: 1,
    system:
      "Responds with concisely written pull request descriptions following the template",
    messages: [
      {
        role: "user",
        content: [
          {
            type: "text",
            text:
              `Please write a pull request description based on the commit using the pull request template.

              Here is the last commit:
              ${lastCommit}
              Here is the template:
              ${template}
            `,
          },
        ],
      },
    ],
  });

  const block = msg.content[0];
  if (block.type === "text") {
    // Check if EDITOR environment variable is set
    const editor = Deno.env.get("EDITOR");

    if (editor) {
      // Create a temporary file with the PR description
      const tempFile = await Deno.makeTempFile({ suffix: ".md" });
      await Deno.writeFile(tempFile, textEncoder.encode(block.text));

      // Open the file with the user's editor
      const process = new Deno.Command(editor, {
        args: [tempFile],
        stdin: "inherit",
        stdout: "inherit",
        stderr: "inherit",
      });

      const status = await process.output();

      // Read the potentially edited content back
      // const editedContent = await Deno.readFile(tempFile);
      // await Deno.writeFile("pr-description.md", editedContent);

      // Clean up the temp file
      await Deno.remove(tempFile);
    } else {
      // Fall back to direct file writing if no editor is set
      Deno.writeFile(
        "genearted-pr-description.md",
        textEncoder.encode(block.text),
      );
    }
  } else {
    console.error("Unexpected message type:", block.type);
  }
}

export { _mockAnthropic };
if (!isTestMode) {
  main();
}
