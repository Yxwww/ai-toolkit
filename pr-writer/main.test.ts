// ai-toolkit/pr-writer/main.test.ts
import { assertEquals, assertStringIncludes } from "jsr:@std/assert";
import { assertSnapshot } from "@std/testing/snapshot";
import {
  assertSpyCall,
  assertSpyCalls,
  returnsNext,
  spy,
  stub,
} from "jsr:@std/testing/mock";
import { _mockAnthropic, main } from "./main.ts";
import * as utils from "./utils.ts";

Deno.test({
  name: "should read PR template",
}, async () => {
  const {
    restore,
    readFileStub,
  } = setupMocks();
  try {
    await main();
  } finally {
    restore();
  }

  // Verify readFileStub was called
  assertSpyCalls(readFileStub, 1);
  const call = readFileStub.calls[0];
  assertEquals(call.args[0], ".github/pull_request_template.md");
});

Deno.test({
  name:
    "Command should call git 3 times to get the diff and open edit on the tempFile",
}, async (t) => {
  const {
    restore,
    gitCommandStub,
  } = setupMocks();
  try {
    await main();
  } finally {
    restore();
  }

  await assertSnapshot(t, gitCommandStub.calls);
});

Deno.test({
  name: "Should call anthropic api with diff and message",
}, async (t) => {
  const {
    restore,
    mockAnthropicClient,
  } = setupMocks();
  try {
    await main();
  } finally {
    restore();
  }

  await assertSnapshot(t, mockAnthropicClient.messages.create.calls);
});

// Mock response for Anthropic API
const mockPrDescription = `## PR Description
Implemented awesome feature

## Changes Made
- Added new component
- Updated documentation

## Testing
Manual testing completed`;

const _internal = {
  readPrTemplate: utils.readPrTemplate,
  readLastCommit: utils.readLastCommit,
};

// Common test setup
const setupMocks = () => {
  const mockTemplate = "## Mock PR Template";
  const mockCommit = "feat: add new feature";
  const stubs: { restore: () => void }[] = [];

  const readFileStub = stub(
    Deno,
    "readTextFile",
    returnsNext([Promise.resolve(mockTemplate)]),
  );
  stubs.push(readFileStub);

  const gitCommandStub = stub(
    Deno,
    "Command",
    () => {
      return {
        output: () =>
          Promise.resolve({
            stdout: new TextEncoder().encode(mockCommit),
            stderr: new Uint8Array(0),
          }),
      };
    },
  );
  stubs.push(gitCommandStub);

  const writeFileStub = stub(
    Deno,
    "writeFile",
    () => Promise.resolve(),
  );
  stubs.push(writeFileStub);

  const makeTempFileStub = stub(
    Deno,
    "makeTempFile",
    () => Promise.resolve("tempfile"),
  );
  stubs.push(makeTempFileStub);

  const removeStub = stub(
    Deno,
    "remove",
    () => Promise.resolve(),
  );
  stubs.push(removeStub);

  const mockMessages = {
    create: spy(() =>
      Promise.resolve({
        content: [
          {
            type: "text",
            text: mockPrDescription,
          },
        ],
      })
    ),
  };

  const mockAnthropicClient = {
    messages: mockMessages,
  };

  const originalAnthropicClient = _mockAnthropic(mockAnthropicClient);

  return {
    mockTemplate,
    mockCommit,
    mockMessages,
    writeFileStub,
    originalAnthropicClient,
    readFileStub,
    mockAnthropicClient,
    gitCommandStub,
    restore() {
      stubs.forEach((stub) => stub.restore());
    },
  };
};
