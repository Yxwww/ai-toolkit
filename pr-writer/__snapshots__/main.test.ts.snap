export const snapshot = {};

snapshot[`Command should call git 3 times to get the diff and open edit on the tempFile 1`] = `
[
  {
    args: [
      "git",
      {
        args: [
          "log",
          "-1",
          "--pretty=%B",
        ],
        stderr: "piped",
        stdout: "piped",
      },
    ],
    returned: {
      output: [Function: output],
    },
    self: stub {},
  },
  {
    args: [
      "git",
      {
        args: [
          "diff",
          "HEAD^",
          "HEAD",
          "--stat",
        ],
        stderr: "piped",
        stdout: "piped",
      },
    ],
    returned: {
      output: [Function: output],
    },
    self: stub {},
  },
  {
    args: [
      "git",
      {
        args: [
          "show",
          "--pretty=format:",
          "--patch",
        ],
        stderr: "piped",
        stdout: "piped",
      },
    ],
    returned: {
      output: [Function: output],
    },
    self: stub {},
  },
  {
    args: [
      "nvim",
      {
        args: [
          "tempfile",
        ],
        stderr: "inherit",
        stdin: "inherit",
        stdout: "inherit",
      },
    ],
    returned: {
      output: [Function: output],
    },
    self: stub {},
  },
]
`;

snapshot[`Should call anthropic api with diff and message 1`] = `
<ref *1> [
  {
    args: [
      {
        max_tokens: 1000,
        messages: [
          {
            content: [
              {
                text: "Please write a pull request description based on the commit using the pull request template.

              Here is the last commit:
              Commit Message:
feat: add new feature

Changes:
feat: add new feature

Code Changes:
feat: add new feature
              Here is the template:
              ## Mock PR Template
            ",
                type: "text",
              },
            ],
            role: "user",
          },
        ],
        model: "claude-3-7-sonnet-20250219",
        system: "Responds with concisely written pull request descriptions following the template",
        temperature: 1,
      },
    ],
    returned: Promise {
      {
        content: [
          {
            text: "## PR Description
Implemented awesome feature

## Changes Made
- Added new component
- Updated documentation

## Testing
Manual testing completed",
            type: "text",
          },
        ],
      },
    },
    self: {
      create: [Function: spy] {
        calls: [Circular *1],
        original: [Function (anonymous)],
        restore: [Function: value],
        restored: [Getter],
      },
    },
  },
]
`;
